app_legacy/Helpers/database/player-game.php

```diff
@@ -91,60 +91,193 @@ function getGameRankAndScore(int $gameID, string $username): array
     return legacyDbFetchAll($query, ['username' => $username])->toArray();
 }
 
-function getUserProgress(string $user, string $gameIDsCSV): array
+function getUserProgress(string $user, array $gameIDs, int $numRecentAchievements = -1, bool $withGameInfo = false): array
 {
-    if (empty($gameIDsCSV) || !isValidUsername($user)) {
-        return [];
-    }
+    $libraryOut = [];
+
+    $awardedData = [];
+    $gameInfo = [];
+    $unlockedAchievements = [];
+    $lockedAchievements = [];
+
+    foreach ($gameIDs as $gameID) {
+        $numAchievements = getGameMetadata($gameID, $user, $achievementData, $gameData);
+
+        $possibleScore = 0;
+        $numAchieved = 0;
+        $scoreAchieved = 0;
+        $numAchievedHardcore = 0;
+        $scoreAchievedHardcore = 0;
+
+        foreach ($achievementData as $achievement) {
+            $points = $achievement['Points'];
+            $possibleScore += $points;
+
+            $dateEarned = $achievement['DateEarned'] ?? null;
+            $dateEarnedHardcore = $achievement['DateEarnedHardcore'] ?? null;
+
+            if ($dateEarned !== null) {
+                $numAchieved++;
+                $scoreAchieved += $points;
+            }
 
-    // Create null entries so that we pass 'something' back.
-    $gameIDsArray = explode(',', $gameIDsCSV);
-    $gameIDs = [];
-    foreach ($gameIDsArray as $gameID) {
-        $gameID = (int) $gameID;
-        $dataOut[$gameID]['NumPossibleAchievements'] = 0;
-        $dataOut[$gameID]['PossibleScore'] = 0;
-        $dataOut[$gameID]['NumAchieved'] = 0;
-        $dataOut[$gameID]['ScoreAchieved'] = 0;
-        $dataOut[$gameID]['NumAchievedHardcore'] = 0;
-        $dataOut[$gameID]['ScoreAchievedHardcore'] = 0;
-        $gameIDs[] = $gameID;
+            if ($dateEarnedHardcore !== null) {
+                $numAchievedHardcore++;
+                $scoreAchievedHardcore += $points;
+            }
+
+            if ($numRecentAchievements >= 0) {
+                if ($dateEarnedHardcore !== null) {
+                    $unlockedAchievements[] = [
+                        'Achievement' => $achievement,
+                        'When' => $dateEarnedHardcore,
+                        'Hardcore' => 1,
+                        'Game' => $gameData,
+                    ];
+                } elseif ($dateEarned !== null) {
+                    $unlockedAchievements[] = [
+                        'Achievement' => $achievement,
+                        'When' => $dateEarned,
+                        'Hardcore' => 0,
+                        'Game' => $gameData,
+                    ];
+                } else {
+                    $lockedAchievements[] = [
+                        'Achievement' => $achievement,
+                        'Game' => $gameData,
+                    ];
```

Comment:
tracking when achievements were unlocked was handled by the `getUsersRecentAwardedForGames` function.